Ipa interface overview (#1)
KevinUTAT/oneAPI-samples_wenkaixu_fork: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/IPAuthoring_interfaces_overview/handshake-control/Readme.md

> # Handshake-control

Reviewer: ```suggestion
# Streaming Invocation and Conduit Arguments
```

This example demonstrate a streaming pipelined kernel with conduit data interfaces and implementing a simple vector addition.

> This design will show how to specify streaming invocation interface through kernel properties and how to specify conduit interface using `annotated_ptr`.

Reviewer: ```suggestion
This implementation uses a streaming invocation interface and conduits to pass the input pointer arguments instead of the CSR. It uses a single, shared memory-mapped host data interface.

This design will show how to specify conduit interfaces using `annotated_ptr` and `annotated_arg`.
```

Also, please include a figure from here (you may need to customize... let me know if you can't find one in here
https://intel-my.sharepoint.com/personal/paul_white_intel_com/Documents/Documentation/HLS%20Docs/kernel-top-levels.vsdx?web=1

## Invocation Interface - Streaming pipelined interface

> In this design, the `SimpleVAdd` kernel is using streaming pipelined invocation interface. This is explicitly specified though kernel properties. 

Reviewer: ```suggestion
In this design, the `SimpleVAdd` kernel uses a streaming pipelined invocation interface. This is explicitly specified using kernel properties. 
```

> To use kernel properties, include either of the following header files:
> ```cpp
> #include <sycl/ext/intel/fpga_extensions.hpp>
> ```
> Or
> ```cpp
> #include <sycl/ext/intel/experimental/fpga_kernel_properties.hpp>
> ```

Reviewer: ```suggestion
```

if it's optional and fpga_extensions.hpp works, then don't make it complicated and just rely on fpga_extensions (which the user is already lincluding anyway)

> In this design following member functions is added to the kernel functor to specify kernel properties:
> ```cpp
> auto get(sycl::ext::oneapi::experimental::properties_tag) {
>     return sycl::ext::oneapi::experimental::properties{
>         sycl::ext::intel::experimental::
>             streaming_interface_accept_downstream_stall,
>         sycl::ext::intel::experimental::pipelined<1>};
>   }
> ```

Reviewer: ```suggestion
In this design, the following member function is added to the kernel functor to specify kernel properties:
auto get(sycl::ext::oneapi::experimental::properties_tag) {
    return sycl::ext::oneapi::experimental::properties{
        sycl::ext::intel::experimental::
            streaming_interface<>,
        sycl::ext::intel::experimental::pipelined<1>};
  }
```

as of the latest 2024.0, you can use `streaming_interface<>` instead of `streaming_interface_...blahblahblah` which is more concise and easier to read. `streaming_interface<>` defaults to `streaming_interface_downstream_stall`

The property `sycl::ext::intel::experimental::streaming_interface_accept_downstream_stall` specify the streaming invocation interface with downstream ready_in interface for back pressuring. You can choose to remove the downstream ready_in interface by using `sycl::ext::intel::experimental::streaming_interface_remove_downstream_stall` instead.

> The property `sycl::ext::intel::experimental::pipelined<1>` specify that this streaming interface is pipelined with an II of 1. You can also pass in other template parameter for this property:

Reviewer: ```suggestion
The property `sycl::ext::intel::experimental::pipelined<1>` specifies that this streaming interface is pipelined with an II of 1. Other valid parameterizations are:
```

- **-1**: Pipeline the kernel, and automatically infer lowest possible II at target fMAX.
- **0**: Do not pipeline the kernel.
- **N (N> 0)**: Pipeline the kernel, and force the II of the kernel to be N.

> > **Note**: `sycl::ext::intel::experimental::pipelined<>` property only supports kernel with streaming interface.

Reviewer: ```suggestion
> **Note**: `sycl::ext::intel::experimental::pipelined<>` property only supports kernels with a streaming invocation interface.
```

## Data Interface

> In this design, all data interfaces (`A_in`, `B_in`, `C_out`, `len`) are implemented as conduits. In kernel with streaming invocation interface, all unannotated argument will be implemented as conduits by default. In this design, they are explicitly specified as conduits through `annotated_ptr` with property `sycl::ext::oneapi::experimental::conduit`.

Reviewer: ```suggestion
In this design, all data interfaces (`A_in`, `B_in`, `C_out`, `len`) are implemented as conduits. In kernels with a streaming invocation interface, all unannotated arguments will be implemented as conduits by default. In this design, they are explicitly specified as conduits to demonstrate the `annotated_ptr` and `annotated_arg` wrappers with property `sycl::ext::oneapi::experimental::conduit`.
```
